pandas/core/reshape/pivot.py

```diff
@@ -482,30 +483,34 @@ def _convert_by(by):
 @deprecate_nonkeyword_arguments(version=None, allowed_args=["data"])
 def pivot(
     data: DataFrame,
-    index: IndexLabel | None = None,
-    columns: IndexLabel | None = None,
-    values: IndexLabel | None = None,
+    index: IndexLabel | lib.NoDefault = lib.NoDefault,
+    columns: IndexLabel | lib.NoDefault = lib.NoDefault,
+    values: IndexLabel | lib.NoDefault = lib.NoDefault,
 ) -> DataFrame:
-    if columns is None:
+    if columns is lib.NoDefault:
         raise TypeError("pivot() missing 1 required argument: 'columns'")
 
     columns_listlike = com.convert_to_list_like(columns)
 
+    data.index.names = [
+        name if name is not None else lib.NoDefault for name in data.index.names
```

Comment:
So names is temporarily set to `lib.NoDefault` here so it could be set to `None` later in `result`?

Reply:
Yep correct. If we keep None and columns is None too, we will end up with 2 MultiIndex levels named None as a temporary results, which causes all kinds of trouble when doing MultiIndex operations and causes multiple failures when doing the pivot. Changing default MultiIndex names to lib.NoDefault is the only thing that would avoid this, but this is quite an invasive change where I am not sure if it would be a good idea at all. hence this workaround here.

Reply:
Added a comment to explain the intent

Reply:
Thanks. Yeah looks like `lib.NoDefault` is hashable too so it should fit the properties of a MultiIndex name.

Alternatively could also use a uuid temp name, but I think `lib.NoDefault` should hopefully be okay

Reply:
Yep lib.NoDefault works great as far as I can see. It's not exposed to the user, correct?

Reply:
Correct.

Reply:
I think we are good then, no risk of collisions